Only define Docker pkg tests if Docker is available

pugnascotia opened the pull request:
Closes #47639, and unmutes tests that were muted in b958467.

The packaging tests are defined dynamically in `DistroTestPlugin`, which includes Docker image tests. The plugin didn't allow for the possibility that Docker isn't installed or is otherwise unavailable when defining the packaging tests, which caused failures when executing on e.g. CentOS 6 or Oracle Linux.

This PR changes the plugin to check for Docker before defining the Docker packaging tests. As part of this, the Docker-checking code in `BuildPlugin` was extracted into a `Docker` utility class, so that it can be called from more than one place.

I'm not particularly proud of the `Docker.java` implementation, it feels a little clumsy, so I'd welcome any thoughts here.

pugnascotia committed:
Revert "Mute docker packaging tests (DockerTests)"

This reverts commit b9584672a460e6b2fcff47ecdd72e856bafa7dc2.

pugnascotia committed:
Only define Docker pkg tests where appropriate

The Docker packaging tests were being defined irrespective of whether
Docker was actually available in the current environment. Change this to
check that Docker is available and vaguely functional before defining
the test tasks.

As part of this, define a seperate utility class for checking Docker,
and call that instead of defining checks in-line in BuildPlugin.groovy

pugnascotia committed:
Fix lint error

pugnascotia committed:
Docs and refactoring

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Build)

pugnascotia committed:
Enable IPv4 forwarding on ubuntu in 16.04

pugnascotia committed:
Use host networking when building docker images

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/DockerUtils.java:
```diff
@@ -0,0 +1,228 @@
+
+            return new Result(process.exitValue(), stdout, stderr);
+        } catch (Exception e) {
+            return new Result(-1, "", e.getMessage());
```
This doesn't seem right. If we are getting some unexpected exception we shouldn't just return a `Result` as though this was what was returned from `docker`. We should rethrow here.

pugnascotia replied:
Exceptions now propagate.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/DockerUtils.java:
```diff
@@ -0,0 +1,228 @@
+     * while running the command, or the process was killed after reaching the 10s timeout,
+     * then the exit code will be -1.
+     */
+    private static Result runCommand(String... args) {
```
This command running stuff seems overly complex to support generic use, yet it's embedded here as an inner-class. I feel like we should either a) just make this a lot simpler, just to solve the "what docker version am I using use-case" or b) rip it out so it can be used elsewhere.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/DockerUtils.java:
```diff
@@ -0,0 +1,228 @@
+     * @return true if the version is high enough
+     */
+    static boolean checkVersion(String dockerVersion) {
+        final String[] majorMinor = dockerVersion.split("\\.");
```
I think we could use `org.elasticsearch.gradle.Version` here, which would allow us to model the minimum required version as a `static Version` constant in this class rather than hard-code it in this method implementation.

pugnascotia replied:
Ah, I didn't know about that class, that's a lot better.

mark-vieira commented on distribution/docker/build.gradle:
```diff
@@ -162,7 +162,7 @@ void addBuildDockerImage(final boolean oss) {
     }
     executable 'docker'
-    final List<String> dockerArgs = ['build', files(oss), '--pull', '--no-cache']
+    final List<String> dockerArgs = ['build', files(oss), '--network', 'host', '--pull', '--no-cache']
```
Why are we switching to host networking here? This has different semantics on a Linux machine with "native" docker vs Docker-for-desktop type scenario, like on OSX where the docker engine is running inside a VM.

pugnascotia replied:
I think that's a mistake, I've removed it.

pugnascotia committed:
Address review feedback

pugnascotia commented on buildSrc/src/main/java/org/elasticsearch/gradle/DockerUtils.java:
```diff
@@ -0,0 +1,228 @@
+        final Optional<String> dockerBinary = getDockerPath();
+
+        if (dockerBinary.isEmpty()) {
+            return dockerAvailability;
```
Thanks for the detailed feedback @mark-vieira! I've reworked how `DockerAvailability` is built, it looks much better IMO.

pugnascotia commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy:
```diff
@@ -209,77 +209,56 @@ class BuildPlugin implements Plugin<Project> {
                 final List<String> tasks = taskGraph.allTasks.intersect(ext.get('requiresDocker') as List<Task>).collect { "  ${it.path}".toString()}
+
                 if (tasks.isEmpty() == false) {
+                    final DockerUtils.DockerAvailability dockerAvailability = getDockerAvailability()
```
Moved.

pugnascotia commented on buildSrc/src/main/java/org/elasticsearch/gradle/DockerUtils.java:
```diff
@@ -0,0 +1,228 @@
+package org.elasticsearch.gradle;
```
Moved.

pugnascotia commented on buildSrc/src/main/java/org/elasticsearch/gradle/DockerUtils.java:
```diff
@@ -0,0 +1,228 @@
+        dockerAvailability.path = dockerBinary.get();
+
+        // Since we use a multi-stage Docker build, check the Docker version since 17.05
+        final Result versionResult = runCommand(dockerAvailability.path, "--version");
```
I did wonder about doing just that, but shied away as the `version` sub-command didn't exist in earlier versions, but I don't think that's a realistic situation that we'll encounter. Switching to `docker version` cleans things up a lot.

atorok commented:
This approach is a bit fragile. On some platforms we always want to have docker and we always want to test for it. Unfortunately Jenkins doesn't make this easy for matrix jobs, but  since we are in more control of  the environment, I'm wondering if we should make this specific, like parsing `/etc/os-release`  or `/etc/issue` and only allow docker to be missing on specific platforms where we know it's not in the images ? 

There's  a similar situation with the test fixtures plugin that skips tests if docker compose is not available but  that's different in that those types of tests should be able to run anywhere ( part of ./gradlew check ) and we didn't want to make docker compose mandatory.  I'we been meaning to add a flag that would force docker-compose to be available and use it in CI. 

My concern is that docker install might be skipped in the image and we'll never notice that we are not running these test.s

pugnascotia commented:
@atorok Right now, the precense or absence of Docker within Vagrant depends on whether we installed Docker via the `Vagrantfile`. However, that doesn't actually answer question: is Docker supposed to be installed in this environment? We could maintain a list / matrix somewhere that specifies where Docker is supposed to be installed, and use that to drive both `Vagrantfile` and `DistroTestPlugin`. What do you think?

pugnascotia commented:
Actually, looking at how the last CI run failed, I think we do need a list of where Docker is and isn't supposed to be installed, because it looks like the CI worker has Docker and so Gradle happily declared the Docker tests tasks, which it then tried to run on `centos-6`, which failed. I'll need to think again here.

atorok commented:
`centos-6` doesn't AFAIK have docker.  The failure in the PR check is infra related 
```
14:40:30 ERROR: Error cloning remote repo 'origin'
```

I think we should assume docker by default, and have a list of the few platforms where it's ok to be missing. 

While at it, could you add a check that runs something like `docker ps`  and also integrate this with `TestFixturesPlugin.dockerComposeSupported` ? 
I think the docker compose stuff should go into the same utility class and have docker supported as a pre condition.  
The reason for `docker ps` has to do with test fixtures. We ran into situations where docker is installed but the user doesn't have permission to it.

pugnascotia committed:
Exclude Docker distro tests per VM

Introduce a VM exclude list, so that the docker packaging tests are run
on all VMs other than those listed. Previously the code checked that
whether Docker was installed, but that wasn't sufficient becasue the
check would happen on the CI host, instead of in the VM.

pugnascotia committed:
Handle Docker version numbers

The Version class is very strict about what suffices it allows after the
patch number, which meant it rejected Docker versions such as
`19.03.2-ce`. Introduce a "relaxed" parsing mode that allow any style of
suffix.

pugnascotia commented:
@elasticmachine run elasticsearch-ci/bwc

pugnascotia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix-docker-tests-on-gcp

pugnascotia committed:
Only run Docker tests on some host OSs

Separately to running packaging tests in Vagrant, we also run packaging
tests through CI on a variety of operating systems. Change how the test
tasks are generated so that the local Docker packaging tests are only
defined if the host OS is supported.

pugnascotia committed:
Make /etc/os-release parsing better

pugnascotia committed:
Try to fix Docker task generation

pugnascotia committed:
Add a debug stmt, figure out debian-8 issue

pugnascotia commented:
@elasticmachine run elasticsearch-ci/packaging

pugnascotia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix-docker-tests-on-gcp

pugnascotia commented:
@elasticmachine test this please

pugnascotia committed:
Fix how we use data from /etc/os-release

pugnascotia committed:
Fix checkstyle

pugnascotia commented:
@elasticmachine run elasticsearch-ci/bwc
@elasticmachine run elasticsearch-ci/default-distro
@elasticmachine run elasticsearch-ci/packaging-matrix

pugnascotia committed:
Trying to diagnose CI failures

pugnascotia committed:
Still trying to diagnose failures on CI

pugnascotia committed:
Please, just print my debug info

pugnascotia committed:
Please, just print my debug info

pugnascotia committed:
Still trying to debug

pugnascotia committed:
Only define docker distros when Docker is available

pugnascotia commented:
@elasticmachine run elasticsearch-ci/packaging-matrix

pugnascotia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix-docker-tests-on-gcp

pugnascotia committed:
Use host networking when building docker images

pugnascotia commented:
@elasticmachine run elasticsearch-ci/packaging-matrix

pugnascotia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix-docker-tests-on-gcp

pugnascotia commented:
@elasticmachine run elasticsearch-ci/1

pugnascotia committed:
Revert docker network change

pugnascotia commented:
@elasticmachine run elasticsearch-ci/packaging-matrix

pugnascotia commented:
Latest status: I've resolved the `centos` etc failures, and the Docker image mysteriously seems to be building on Ubuntu now, so the only reliable failure is `debian-10`, which is nothing to do with this PR. Debian aside, the last `packaging-matrix` run only failed due to some transient errors on Ubuntu 1604. I'm hopeful we can overlook Debian and think about merging....?

pugnascotia commented:
Now blocked by https://github.com/elastic/infra/issues/15654

pugnascotia committed:
Workaround infra issue 15654

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into fix-docker-tests-on-gcp

pugnascotia committed:
Increase docker pkg test timeouts

Slowness in CI can completely ruin a Docker test run, since the current
timeouts are a bit tight.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into fix-docker-tests-on-gcp

pugnascotia commented:
@elasticmachine run elasticsearch-ci/packaging-matrix because I'd like to test the latest commit, thank you very much.

pugnascotia commented:
@elasticmachine run elasticsearch-ci/packaging-matrix again, with feeling

pugnascotia committed:
Tiny change to try and get CI to work

pugnascotia committed:
Docker test fixes following JVM options work

pugnascotia commented:
PR is green! Please review!

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/tool/DockerUtils.java:
```diff
@@ -0,0 +1,242 @@
+            throwDockerRequiredException("Something went wrong while checking Docker's availability", e);
+        }
+
+        if (availability.isAvailable == true) {
```
Although the `== false` is a convention we use, I'm not sure `boolean == true` is conventional in the ES project. I think we can just simplify this to `if (availability.isAvailable)`.

mark-vieira commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java:
```diff
@@ -65,6 +70,40 @@
+     * listed here. These values correspond to the sub-project names, which
+     * correspond to the directory names.
+     */
+    private static final List<String> DOCKER_VM_EXCLUDE_LIST = List.of(
```
I'd prefer we not hard-code this kind of thing and instead have the projects themselves _declare_ this kind of thing.
Basically, the project should tell the plugin about this limitation, not the other way around.

I'd do something like:
1. Register an extension (or extra property) on each subproject to allow it to declare it supports Docker packaging tests.
2. Disable the docker packaging tests tasks via an `onlyIf` condition that looks for that extension property value.

mark-vieira commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java:
```diff
@@ -65,6 +70,40 @@
+     * of whether to run the Docker tests is answered by {@link
+     * #shouldRunDockerTests()}.
+     */
+    private static final List<String> DOCKER_LINUX_INCLUDE_LIST = List.of(
```
My comment above applies here as well.

Also, isn't this essentially the inverse of the above list?

mark-vieira commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java:
```diff
@@ -81,6 +120,8 @@
     private static final String IN_VM_SYSPROP = "tests.inVM";
     private static final String DISTRIBUTION_SYSPROP = "tests.distribution";
 
+    private static final boolean dockerTestEnabled = shouldRunDockerTests();
```
Can we try and defer this until we actually need it? Executing `shouldRunDockerTests()` hits the filesystem and if possible we want to avoid any and all configuration time I/O from statically initializing these kinds of things.

mark-vieira commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java:
```diff
@@ -326,12 +377,20 @@ public String toString() {
+                        if (dockerTestEnabled == false) {
+                            skip = true;
+                        } else if (bundledJdk == false) {
+                            // We should never add a Docker distro with bundledJdk == false
```
I think we could improve this comment. Something like:

```
// All our Docker images include a bundled JDK so it doesn't make sense to test without one
```

mark-vieira commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java:
```diff
@@ -405,4 +465,80 @@ private static String destructiveDistroTestTaskName(ElasticsearchDistribution di
+                return false;
+
+            case MAC:
+                // Assume that Docker for Mac is installed
```
Do we _need_ to assume? Couldn't we leverage `DockerAvailability` here to make this a bit more robust? Also, should we even explicitly handle MAC since it's not an OS that we run _any_ packaging tests on.

mark-vieira commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java:
```diff
@@ -405,4 +465,80 @@ private static String destructiveDistroTestTaskName(ElasticsearchDistribution di
+
+                    final boolean contains = DOCKER_LINUX_INCLUDE_LIST.contains(id);
+
+                    logger.debug("Linux OS id [" + id + "] is " + (contains ? "" : "not ") + "present in the include list");
```
Here's the thing about debug logging... it's never on in CI, and if you turn it on locally, the amount of shit that gets logged pretty much drowns out anything you might otherwise be looking for. If this is useful for local debugging I'd bump it to `info`, otherwise include in an exception message where appropriate.

mark-vieira commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java:
```diff
@@ -405,4 +465,80 @@ private static String destructiveDistroTestTaskName(ElasticsearchDistribution di
+
+                    return contains;
+                } else {
+                    logger.debug("/etc/os-release does not exist!");
```
As mentioned above, if this is a unexpected case we shouldn't encounter we should make this more apparent or throw an error. Logging this at debug is effectively invisible.

mark-vieira reviewed (changes requested):
Left some feedback

pugnascotia committed:
Simplify how Docker test tasks are defined

Rather than using some complicate /etc/os-release parsing to determine
whether to define docker packaging test tasks, re-use the methods in
DockerUtils and just check if Docker can be used.

Also, when defining the VM tasks, rely on properties in the qa/os
sub-projects to control whether to run Docker tests inside the VMs.

pugnascotia commented:
@mark-vieira it may be helpful if I outline a few reasons why the code is doing what it's doing in this PR:

   * The packaging test tasks have dependencies on the artificts that will be tested. `DistributionDownloadPlugin` works out whether these should be pulled down from a repo, or built in the current checkout
   * The Docker packaging tests require a Docker image, and the test dependencies ensure that it's possible to build and load the Docker image
   * ...but the `:distribution:docker` project makes a call to `requireDocker`, which checks whether Docker is available while configuring the build
   * So the docker packaging tests have to work out whether Docker is available before configuring Docker packaging tests, or the build fails. That's what all the include / exclude list stuff is all about, whether to configure the docker tests.

I just did an experiment using some `onlyIf` predicates, but it looks like if a task is skipped in this way, its dependencies are still executed, so I'd need to put an `onlyIf` on all the Docker-related tasks, which feels wrong.

All that said, I've pushed some simplificiations (hopefully), and more feedback would be great.

pugnascotia commented:
Tagging @atorok as I don't know whether what I've done with the `:qa:os` subprojects is the right approach or what @mark-vieira had in mind.

atorok commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java:
```diff
@@ -128,7 +136,18 @@ public void apply(Project project) {
+                        // Only VM sub-projects that are specifically opted-in to testing Docker should
+                        // have the Docker task added as a dependency. The shouldTestDocker property could
+                        // be null, hence we use Boolean.TRUE.equals()
+                        boolean shouldExecute = distribution.getType() != Type.DOCKER
```
The typical way to go about this is to register an Gradle extension on the project where the project could provide additional configuration, like weather or not to test on docker. 

Aslo, I think we should specifically opt-put  instead of opt in as most platforms support it and future platforms are expected to support it.

mark-vieira replied:
Not sure what earlier comment you are referring to @atorok. 

Also, I think this is going to have to go in an `afterevaluate { }` block to ensure that property has been evaluated.

atorok replied:
I was referring to this one: https://github.com/elastic/elasticsearch/pull/47640#discussion_r340652049 didn't realize GH will put it after not before. 

My point is that we don't need to bother with docker support for the subprojects for when docker is not available in the Vagrant VM ( which is when these are actually used ) . The tests are skipped when there's no docker and there is no need to build an image inside the VM, so we're all good. Not having docker on the host that runs vagrant is ok to error out.

atorok commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java:
```diff
@@ -405,4 +425,16 @@ private static String destructiveDistroTestTaskName(ElasticsearchDistribution di
+    private static boolean isDockerAvailable() {
+        try {
+            return getDockerAvailability().isAvailable;
+        } catch (Exception e) {
```
it's ok to fail in the case of some unexpected error. A failed build Today is better than realizing after 2 months that we were not testing on Docker when we should have

pugnascotia commented:
I've just thought of a drawback of always relying on detecting whether Docker is available. What if we broke Docker on the GCP images, and so suddenly stopped testing Docker? The build wouldn’t fail unless it knew that it ought to have Docker available on a given host. I don't have a list of projects for the GCP hosts either. That's why I ended up with the lists of host types in the first place. And in fact, [Alpar said as much a few weeks ago](https://github.com/elastic/elasticsearch/pull/47640#issuecomment-539972802).

pugnascotia commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java:
```diff
@@ -99,10 +103,14 @@ public void apply(Project project) {
         TaskProvider<Copy> copyUpgradeTask = configureCopyUpgradeTask(project, upgradeVersion, upgradeDir);
         TaskProvider<Copy> copyPluginsTask = configureCopyPluginsTask(project, pluginsDir);
 
+        final boolean dockerAvailable = isDockerAvailable();
```
This code is part of the configuration of the `destructiveDistroTest` task though - I'm not expecting anyone to explicitly run the Docker tests during normal operations, it should just be part of the overall suite of distro tests. It's only tricky because we don't have Docker everywhere.

atorok replied:
you're right, thinking about it again, you're right, it's fine to have it here.

atorok commented:
> > I think the easiest way to accomplish this is to skip configuring the distribution altogether in this case, similar to how we deal with docker rest tests.
> 
> I'm not sure what you mean here. CI runs the `destructivePackagingTest` task, and relies on task dependencies to get everything done. The Docker tests do the same as the other tests, in that they depend on a distribution task to build the Docker image, so that it can be tested.

What I meant is that we could avoid adding those task dependencies in the first place. 
That's already what you are doing in the PR,  I was just trying to highlight it.

pugnascotia committed:
Reimplement Docker tests exclude list

Relying on whether Docker is available on a host to drive running the
Docker tests is not sufficient, because we expect and requrie Docker to
be available on some OS versions. Revert to maintaining a list, and
externalise it to `.ci/dockerOnLinuxExclusions` in the repository root.

atorok commented:
> I'm not keen on doing an refactoring in `TestFixturesPlugin.dockerComposeSupported()` in this PR. I understand that it's related, but it feels outside the scope of what I'm trying to do here.

I'm perfectly fine with doing it in a follow up. This is large as it is.

pugnascotia committed:
Remove sles-15 from .ci/dockerOnLinuxExclusions

pugnascotia commented on buildSrc/src/main/java/org/elasticsearch/gradle/tool/DockerUtils.java:
```diff
@@ -0,0 +1,242 @@
+     * @param tasks the tasks that require Docker
+     * @throws GradleException if Docker is not available. The exception message gives the reason.
+     */
+    public static void assertDockerIsAvailable(List<String> tasks) {
```
This method is used in `BuildPlugin.groovy`.

atorok replied:
Sorry, I misplaced this comment, I meant it for `getDockerAvailability`

pugnascotia replied:
Ah, I see - yes, this method used to be used externally, but then things were refactored. I'll make it `private`.

pugnascotia commented:
@elasticmachine run elasticsearch-ci/packaging-matrix

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/tool/DockerUtils.java:
```diff
@@ -0,0 +1,242 @@
+     * while running the command, or the process was killed after reaching the 10s timeout,
+     * then the exit code will be -1.
+     */
+    private static Result runCommand(String... args) throws Exception {
```
A new `ExecOperations` service is available in Gradle 6.0 so we'd have to wait until that is release to leverage it.

atorok replied:
Do you think it makes sense to pass in a `project` instance in the mean time and call `project.exec` ?

mark-vieira commented:
> So the docker packaging tests have to work out whether Docker is available before configuring Docker packaging tests, or the build fails. That's what all the include / exclude list stuff is all about, whether to configure the docker tests.

This make sense. Thanks for the explanation @pugnascotia.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/tool/DockerUtils.java:
```diff
@@ -0,0 +1,242 @@
+            lastResult = runCommand(dockerPath, "version", "--format", "{{.Server.Version}}");
+
+            if (lastResult.isSuccess() == true) {
+                version = Version.fromString(lastResult.stdout.trim(), Version.Mode.RELAXED);
```
Ideally then we'd have a generic `Version` class and then an `ElasticsearchVersion` with additional semantics as Alpar describes. We deal with parsing version numbers so often it just make sense to have a utility like this and `Version` as it exists is 99% of what we need.

mark-vieira commented on .ci/dockerOnLinuxExclusions:
```diff
@@ -0,0 +1,13 @@
+# This file specifies the Linux OS versions on which we can't build and
```
Couldn't we infer this from the `shouldTestDocker` information? Is there a need to keep this list? Basically, if there isn't a corresponding `:os` project with `shouldTestDocker = true` for the given OS then don't setup the Docker test tasks.

atorok replied:
We can't. the sub-projects are used only for Vagrant and we need to skip docker on gcp too.  We already ski docker in tests so I don't see the point of `shouldTestDocker`, as we'll always have docker on the host to be able to build the image, and we'll always have to build it for some platform. Not having docker on the vagrant host should error out as it can't be tested at all.

mark-vieira replied:
Gotcha.

atorok replied:
We _do_ have  docker on all CI Vagrant hosts. There's no host running `centos-6` we only run that as vagrant guest. And it's ok to auto-detect docker on Vagrant guests, because we have the GCP tests as the source of truth and we are specific about it  there.

pugnascotia committed:
Fix access modifiers in DockerUtils

pugnascotia committed:
Use project.exec to run commands in DockerUtils

pugnascotia commented:
I've pushed some further changes following review comments.

I feel like we're talking in circles now about some of the finer points of the implementation. Can I suggest merging now as-is, subject to a clear CI run of course, since we'll certinaly iterate more on this over time? That way we at least get some Docker image test coverage in CI again. I know I'll be working more on this area as I have #47573 to finish and merge, and I have more packaging tests to port over from the old `elasticsearch-docker` repo.

pugnascotia committed:
Fix

pugnascotia committed:
Fix JavaDoc ref

pugnascotia committed:
Don't always create a docker distribution

Restore the logic that avoided creating a Docker distribution in
DistroTestPlugin, because that has the effect of building the Docker
image even when we don't want to test it.

pugnascotia committed:
Fix logic

mark-vieira commented:
> I feel like we're talking in circles now about some of the finer points of the implementation. Can I suggest merging now as-is, subject to a clear CI run of course, since we'll certinaly iterate more on this over time?

👍

pugnascotia commented:
@elasticmachine run elasticsearch-ci/2 since it experience a peculiar transient error.

pugnascotia commented:
@elasticmachine run elasticsearch-ci/packaging-matrix as, yet again, the wrong commit was tested.

pugnascotia committed:
Update  .ci/dockerOnLinuxExclusions

atorok commented:
I would really like us to remove `project.ext.shouldTestDocker` before merging, as it's really confusing and unnecessary i.m.h.o

pugnascotia committed:
Expand comment re: configuring Docker tasks for Vagrant

pugnascotia commented:
> I would really like us to remove `project.ext.shouldTestDocker` before merging, as it's really confusing and unnecessary i.m.h.o

@atorok I tried removing this and relying on detecting whether Gradle is running in a VM via `IN_VM_SYSPROP`, but the problem is that the test tasks are defined before we even have a VM running, so detection cannot work here. I'm going to leave `shouldTestDocker`, and maybe we can think of something better in the future.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into fix-docker-tests-on-gcp

pugnascotia committed:
Tweak logging

pugnascotia commented:
@elasticmachine run elasticsearch-ci/packaging-matrix to pick up the right commit

pugnascotia committed:
Add debugging statements

pugnascotia commented:
@elasticmachine run elasticsearch-ci/packaging-matrix to pick up the right commit

pugnascotia committed:
Pointless change to get CI to run the right code

pugnascotia committed:
Various fixes to docker packaging test logic

pugnascotia commented:
@elasticmachine run elasticsearch-ci/packaging-matrix to pick up the right commit

pugnascotia committed:
Pointless change to try and get CI to run the right code

pugnascotia commented:
@elasticmachine run elasticsearch-ci/packaging-matrix to pick up the right commit

pugnascotia commented:
@elasticmachine run elasticsearch-ci/2

pugnascotia commented:
It's green! Green I tell you!

atorok commented:
> > I would really like us to remove `project.ext.shouldTestDocker` before merging, as it's really confusing and unnecessary i.m.h.o
> 
> @atorok I tried removing this and relying on detecting whether Gradle is running in a VM via `IN_VM_SYSPROP`, but the problem is that the test tasks are defined before we even have a VM running, so detection cannot work here. I'm going to leave `shouldTestDocker`, and maybe we can think of something better in the future.

@pugnascotia I don't quite see why this is necessary. AFAIK, When running tests in a VM, we run the tasks in the sub-projects, and when running in the GCP the tasks are attached to the `qa/os` projects,  so these can have different dependencies. I was under the impression that when running in vagrant we do not need any detection at the Gradle level, the tests will just skip if docket is not available on the VM.  With GCP, we 1. make sure docker is available on platforms we expect it to be and 2. make sure we don't pull in docker dependencies on platforms where it's not available.
